core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4628,6 +4647,73 @@ private void analyzeRowFilter(String currentIdentity, Table table, QualifiedObje
             analysis.addRowFilter(table, expression);
         }
 
+        private void analyzeCheckConstraint(Table table, QualifiedObjectName name, Scope scope, ViewExpression filter)
+        {
+            Expression expression;
+            try {
+                expression = sqlParser.createExpression(filter.getExpression(), createParsingOptions(session));
+            }
+            catch (ParsingException e) {
+                throw new TrinoException(INVALID_ROW_FILTER, extractLocation(table), format("Invalid check constraint for '%s': %s", name, e.getErrorMessage()), e);
+            }
+
+            verifyNoAggregateWindowOrGroupingFunctions(session, metadata, expression, format("Check constraint for '%s'", name));
+            verifyNoSubqueryForCheckConstraint(expression);
+
+            ExpressionAnalysis expressionAnalysis = getExpressionAnalysis(table, name, scope, filter, expression, "Invalid check constraint");
+
+            // Ensure that the expression doesn't contain non-deterministic functions. This should be "retrospectively deterministic" per SQL standard.
+            if (!DeterminismEvaluator.isDeterministic(expression, this::getResolvedFunction)) {
```

Comment:
I saw the discussion on Slack about this. It turns out this is much trickier when functions such as `current_timestamp`, `localtimestamp`, etc are involved, and just checking for non-determinism is not strict enough:

<img width="809" alt="image" src="https://user-images.githubusercontent.com/14387/208753085-a98d8e60-7ba3-4bc1-9849-3854d95098a7.png">

We need to either disallow those functions entirely (for now, and do it as a future improvement), or implement the proper semantics.

Reply:
Added a logic to deny those functions.